invariant/examples/sql_injection/sql_injection_workflow.ipynb

```diff
@@ -0,0 +1,720 @@
+    "        return result\n",
+    "        \n",
+    "    def __call__(self, message: str) -> str:\n",
+    "        query_bundle = QueryBundle(query_str=message)\n",
+    "        response = self._query(query_bundle)\n",
+    "        return response\n",
+    "\n",
+    "sql_query_engine = CallableCustomNLSQLTableQueryEngine(\n",
+    "    sql_database=sql_database,\n",
+    "    tables=[\"albums\", \"tracks\", \"artists\"],\n",
+    "    verbose=True,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "id": "6f474033-6a5b-470c-a36e-3505dbc94c92",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from llama_index.core.tools import FunctionTool\n",
+    "from llama_index.llms.openai import OpenAI\n",
+    "\n",
+    "\n",
+    "tools = [\n",
+    "    FunctionTool.from_defaults(sql_query_engine),\n",
+    "]\n",
+    "\n",
+    "message_logs = []\n",
+    "\n",
+    "validate_prompt_str = \"\"\"\\\n",
+    "You performed a SQL query and retrieved a SQL response. Judge whether the SQL response is enough to answer the user question.\n",
+    "Answer with YES or NO.\n",
+    "\n",
+    "SQL query: {sql_query}\n",
+    "SQL Response: {response}\n",
+    "\n",
+    "Result: \"\"\"\n",
+    "retry_prompt_str = \"\"\"\\\n",
+    "Based on the previous response, generate a new SQL query and call tools to perform it.\n",
+    "Previous response: {output}\n",
+    "SQL query to perform:\"\"\"\n",
+    "\n",
+    "agent = ReActAgent(\n",
+    "    llm=CustomOpenAI(model=\"gpt-4o-mini\", temperature=0.8), tools=tools, validate_prompt=validate_prompt_str, retry_prompt=retry_prompt_str, timeout=120, verbose=True\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "433a7d5f-f757-4bac-86a4-b8e5cb4ac0b1",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Running step new_user_msg\n",
+      "Step new_user_msg produced event PrepEvent\n",
+      "Running step prepare_chat_history\n",
+      "Step prepare_chat_history produced event InputEvent\n",
+      "Running step handle_llm_input\n",
+      "Step handle_llm_input produced event ToolCallEvent\n",
+      "Running step handle_tool_calls\n",
+      "Step handle_tool_calls produced event PrepEvent\n",
+      "Running step prepare_chat_history\n",
+      "Step prepare_chat_history produced event InputEvent\n",
+      "Running step handle_llm_input\n",
+      "Step handle_llm_input produced event PrepEvent\n",
+      "Running step prepare_chat_history\n",
+      "Step prepare_chat_history produced event InputEvent\n",
+      "Running step handle_llm_input\n",
+      "Step handle_llm_input produced event ToolCallEvent\n",
+      "Running step handle_tool_calls\n",
+      "Step handle_tool_calls produced event StopEvent\n"
+     ]
+    }
+   ],
+   "source": [
+    "ret = await agent.run(input=\"Who is the first artist in the table `artists`?\")"
```

Comment:
@mbalunovic I cannot reproduce your error. Below is my result of `pip freeze`.

```
aiohappyeyeballs==2.3.6
aiohttp==3.10.3
aiosignal==1.3.1
annotated-types==0.7.0
anyio==4.4.0
appnope==0.1.4
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
arrow==1.3.0
asttokens==2.4.1
async-lru==2.0.4
attrs==23.2.0
autobahn==23.6.2
Automat==22.10.0
Babel==2.15.0
beautifulsoup4==4.12.3
bleach==6.1.0
boltons==21.0.0
bracex==2.5
certifi==2024.7.4
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
click-option-group==0.5.6
colorama==0.4.6
comm==0.2.2
constantly==23.10.4
cryptography==42.0.8
dataclasses-json==0.6.7
debugpy==1.8.2
decorator==5.1.1
defusedxml==0.7.1
Deprecated==1.2.14
dirtyjson==1.0.8
distro==1.9.0
docker-pycreds==0.4.0
exceptiongroup==1.2.2
executing==2.0.1
face==22.0.0
fastjsonschema==2.20.0
filelock==3.15.4
fqdn==1.5.1
frozenlist==1.4.1
fsspec==2024.6.1
gitdb==4.0.11
GitPython==3.1.43
glom==22.1.0
googleapis-common-protos==1.63.2
greenlet==3.0.3
h11==0.14.0
hkdf==0.0.3
httpcore==1.0.5
httpx==0.27.0
huggingface-hub==0.24.5
humanize==4.10.0
hyperlink==21.0.0
idna==3.7
importlib_metadata==7.1.0
incremental==22.10.0
invariant @ git+https://github.com/invariantlabs-ai/invariant.git@9edf9881bbabdd4461104492074a985043e6ac3b
ipykernel==6.29.5
ipython==8.26.0
isoduration==20.11.0
iterable-io==1.0.0
jedi==0.19.1
Jinja2==3.1.4
jiter==0.5.0
joblib==1.4.2
json5==0.9.25
jsonpickle==3.2.2
jsonpointer==3.0.0
jsonschema==4.23.0
jsonschema-specifications==2023.12.1
jupyter-events==0.10.0
jupyter-lsp==2.2.5
jupyter_client==8.6.2
jupyter_core==5.7.2
jupyter_server==2.14.2
jupyter_server_terminals==0.5.3
jupyterlab==4.2.3
jupyterlab_pygments==0.3.0
jupyterlab_server==2.27.2
lark==1.2.2
llama-cloud==0.0.14
llama-index==0.10.68
llama-index-agent-openai==0.2.9
llama-index-cli==0.1.13
llama-index-core==0.10.68.post1
llama-index-embeddings-huggingface==0.2.3
llama-index-embeddings-openai==0.1.11
llama-index-indices-managed-llama-cloud==0.2.7
llama-index-legacy==0.9.48.post3
llama-index-llms-ollama==0.2.2
llama-index-llms-openai==0.1.29
llama-index-llms-replicate==0.1.3
llama-index-multi-modal-llms-openai==0.1.9
llama-index-program-openai==0.1.7
llama-index-question-gen-openai==0.1.3
llama-index-readers-file==0.1.33
llama-index-readers-llama-parse==0.1.6
llama-parse==0.4.9
magic-wormhole==0.14.0
markdown-it-py==3.0.0
MarkupSafe==2.1.5
marshmallow==3.21.3
matplotlib-inline==0.1.7
mdurl==0.1.2
minijinja==2.0.1
mistune==3.0.2
mpmath==1.3.0
multidict==6.0.5
mypy-extensions==1.0.0
nbclient==0.10.0
nbconvert==7.16.4
nbformat==5.10.4
nest-asyncio==1.6.0
networkx==3.3
nltk==3.8.1
notebook_shim==0.2.4
numpy==1.26.4
ollama==0.3.1
openai==1.40.8
opentelemetry-api==1.25.0
opentelemetry-exporter-otlp-proto-common==1.25.0
opentelemetry-exporter-otlp-proto-http==1.25.0
opentelemetry-instrumentation==0.46b0
opentelemetry-instrumentation-requests==0.46b0
opentelemetry-proto==1.25.0
opentelemetry-sdk==1.25.0
opentelemetry-semantic-conventions==0.46b0
opentelemetry-util-http==0.46b0
overrides==7.7.0
packaging==24.1
pandas==2.2.2
pandocfilters==1.5.1
parso==0.8.4
peewee==3.17.6
pexpect==4.9.0
pillow==10.4.0
platformdirs==4.2.2
prometheus_client==0.20.0
prompt_toolkit==3.0.47
protobuf==4.25.4
psutil==6.0.0
ptyprocess==0.7.0
pure-eval==0.2.2
pyasn1==0.6.0
pyasn1_modules==0.4.0
pycparser==2.22
pydantic==2.8.2
pydantic_core==2.20.1
Pygments==2.18.0
PyNaCl==1.5.0
pyOpenSSL==24.1.0
pypdf==4.3.1
python-dateutil==2.9.0.post0
python-json-logger==2.0.7
pytz==2024.1
pyvis==0.3.2
PyYAML==6.0.1
pyzmq==26.0.3
referencing==0.35.1
regex==2024.7.24
requests==2.32.3
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rich==13.7.1
rpds-py==0.19.0
ruamel.yaml==0.17.40
ruamel.yaml.clib==0.2.8
safetensors==0.4.4
scikit-learn==1.5.1
scipy==1.14.0
semgrep==1.85.0
Send2Trash==1.8.3
sentence-transformers==3.0.1
sentry-sdk==2.10.0
service-identity==24.1.0
setproctitle==1.3.3
setuptools==70.3.0
six==1.16.0
smmap==5.0.1
sniffio==1.3.1
soupsieve==2.5
spake2==0.8
SQLAlchemy==2.0.32
stack-data==0.6.3
striprtf==0.0.26
sympy==1.13.0
tenacity==8.5.0
termcolor==2.4.0
terminado==0.18.1
threadpoolctl==3.5.0
tiktoken==0.7.0
tinycss2==1.3.0
tokenizers==0.19.1
tomli==2.0.1
torch==2.3.1
tornado==6.4.1
tqdm==4.66.4
traitlets==5.14.3
transformers==4.44.0
Twisted==24.3.0
txaio==23.1.1
txtorcon==23.11.0
types-python-dateutil==2.9.0.20240316
typing-inspect==0.9.0
typing_extensions==4.12.2
tzdata==2024.1
uri-template==1.3.0
urllib3==2.2.2
wandb==0.17.4
wcmatch==8.5.2
wcwidth==0.2.13
webcolors==24.6.0
webencodings==0.5.1
websocket-client==1.8.0
wrapt==1.16.0
yarl==1.9.4
zipp==3.20.0
zipstream-ng==1.7.1
zope.interface==6.4.post2
```

Reply:
@mbalunovic I just found out that this is due to different versions of llama-index. It worked for `llama-index==0.10.68` and perhaps they made a recent update and this is now broken.
There are two things we can do here
1. I just change the `pip install llama_index` into `pip install llama-index==0.10.68`.
2. We follow their upstream and debug the problem.

I personally prefer 1 because we cannot keep maintaining this. At some point it will be impossible to catch up. What do you think?

Reply:
Let's try to make it work with the newest version. I think this should be resolved by replacing

```python
from llama_index.legacy.core.llms.types import (
    ChatMessage,
    ChatResponse,
)
```

with

```python
from llama_index.core.base.llms.types import (
    ChatMessage,
    ChatResponse,
)
```

At least with this it works on my machine. The bug was that ChatMessage was invalid type (of the same name) which caused Exception in the Pydantic. Could you verify that it works for you now?